rfcs/45-stageless.md

```diff
@@ -0,0 +1,984 @@
+The checker can point out what's ambiguous before *and* after flattening dependency graphs.
+(i.e. the checker can report that system A conflicts with system set X before reporting the specific pairs of conflicting systems).
+
+We hope system sets will help reduce the cognitive burden here and naturally lead users to patterns with fewer errors.
+Likewise, if users also become responsible for scheduling sets exported by plugins, it will be within their power to resolve any errors.
+
+Still, we have identified follow-up work to improve this.
+
+If you have two ambiguously-ordered system sets that conflict, you probably won't get the behavior you expect if all their systems just get mixed together.
+[RFC #46](https://github.com/bevyengine/rfcs/pull/46) (which spun out of this one) aims to give system sets configurable atomicity to prevent this mixing from happening, although the ambiguity would still exist.
+
+### Why is command application scheduled with an exclusive system?
+
+Several earlier RFCs talked about expressing more exact dependencies, e.g. `B.after_buffers_of(A)`, where such graph edges could hypothetically be used to automatically determine when to apply commands, at a minimal number of points.
+We don't know how to achieve that hypothetical (graph problems are difficult).
+However, if contributors find a solution later, they will be able to build on top of this design.
+
+Therefore, although `apply_system_buffers` might not be the best way to handle this, it will let users decide when commands are applied and ensure other systems aren't running when that happens.
+We think that's good enough for now.
+
+(Maybe instead of `apply_system_buffers` applying the pending commands of *all* systems in the current schedule, we could construct `apply_system_buffers` from a tuple of system (set) labels, such that only those have their commands applied.
+It's debatable whether that would increase or reduce cognitive burden.
+That can be revisited later too.)
+
+### Why did you change states from a stack to a basic FSM? Where is `OnUpdate(X)`?
+
+If this RFC is merged and this design is implemented, we expect most of the problems that encouraged it to vanish.
+So rather than port pieces of the existing API, we think it's better to start fresh with a basic state machine and see what patterns emerge and what their limits are.
+If, after migration, a significant number of users are still turning towards a plugin to recover the lost stack functionality, we can consider upstreaming it again.
+
+A stack-based state machine should be trivial to implement though (i.e. with a loop inside the exclusive system).
+
+As for `OnUpdate(X)`... its interaction with `OnEnter(X)` was non-intuitive.
+Users often just wanted systems to run in their designated spots, but only when a certain state was active. That is much simpler to do now (see example below), so having a system set for this that you can only put in one spot seemed pointless.

```

Comment:
I'm not opposed to this. I think that this is a bit clearer, and the argument for visualization is compelling.

Reply:
I don't think I fully understand your POV yet.

Are you expecting `OnUpdate(X)` to be a system set to be run or simply an "abstract logical group"? 
If you intend to run it, when should it be run?

> enable us to provide standard / predictable orderings between things like `OnEnter(GameState::Playing)` and `OnUpdate(GameState::Playing)`.

Currently, you can't order anything relative to `OnEnter(X)` and `OnExit(X)`. They don't exist within the main app update set (because you can't add multiple instances of a system set, i.e. `OnEnter(X)`, but users may want to have transitions at multiple points in their app update). So they're "detached" and summoned from within `apply_state_transition` instances. 

> From my perspective having one preferred spot is the point.

Is there value in having the engine dictate a standard location for state updates? Control over plugin system scheduling is such a pivotal issue, so I find it a bit strange that state logic would be treated differently.

Users could very easily create `OnUpdate(X)` themselves.
I think having both `OnUpdate(X)` and `state_equals` will cause confusion.

Reply:
I'm not sure if an `OnUpdate` would make things better. You can already create a system set and guard it with `state_equals`. Here, adding what is basically sugar may cause more confusion than help.

Finally, to me the API seems already symmetrical: `OnEnter` and `OnExit` are just for state *transitions*, while the other update systems run just like any other condition guarded system. Also, while there are very good reasons to group systems under a set in a transition, that's not always the case for update ones.

Reply:
You articulated the point about API symmetry much better than I did.

Yeah, state *transitions* are special and have to work this way, "state_equals" does not. They aren't the same, so the lack of symmetry is natural. IMO the old API got this wrong.

Lots of users try to add `on_update` to multiple stages, which doesn't work (bevyengine/bevy#1865). A `state_equals` criteria/condition does work, but I think users don't realize that `on_update` *isn't* that.

Also the unique property `on_update` currently has where it also runs after entering a state confused multiple people (bevyengine/bevy#3293 and bevyengine/bevy#4236).

Reply:
> Yeah, state transitions are special and have to work this way, state_equals does not. They aren't the same, so the lack of symmetry is natural.

This is compelling to me. I'm with @maniwani after reviewing these discussions.

> And the special property on_update has where it also runs after entering a state has confused multiple people (https://github.com/bevyengine/bevy/issues/3293 and https://github.com/bevyengine/bevy/issues/4236). I don't think that property is necessary

I agree here. This property was only important if we're processing many state transitions in a single frame via looping. In rare cases when that behavior is still desired, it's easy enough to manually and explicitly control.

Reply:
> Yeah, state transitions are special and have to work this way, state_equals does not. They aren't the same, so the lack of symmetry is natural.

I agree with this as an argument that symmetry doesn't follow from those aspects of the problem, but I disagree that symmetry isn't natural or desirable within the global context of the system. State transitions still use the "labeled set" api, which is the standard way to address a group of systems. My argument is that "state update logic" should be named / addressable / orderable. The symmetry that follows seems very natural (and nice because of the consistency when it comes to defining logic). 

If you don't buy that "state update logic" should be named / addressable / orderable, then the symmetry argument starts to break down.

> Are you expecting OnUpdate(X) to be something that gets run directly or simply an "abstract logical group"?
If you intend to run it, when should it be run?

I don't think it should be run directly, nor do I think it should "just" be an "abstract logical group". I think it should be a named group of systems that exists within the larger `CoreSystems::Update` set, which you can order the execution of other systems around, most importantly with predictable/standard orderings relative to OnEnter/OnExit.

> So OnEnter(X) and OnExit(X) are "detached" and summoned from within apply_state_transition instances.

Ahh so this is fully an exercise for the user. Not something with a standard place in the schedule. Not yet sure how I feel about that. Seems like a footgun / something easy to forget. But forcing users to be intentional about state transitions _is_ also nice, once they're aware of it.

In a world where we're forcing users to manually define the structure of state logic and transitions in the schedule, it does make less sense to have one standard OnUpdate.

That being said, I think standardizing "state lifecycles" for the common case is "nice". My guess is that 95% of the time, people won't want to / shouldn't need to worry about manually deciding where to apply state changes, where to insert state update logic in the schedule, etc. I think a standard "define the on_enter, on_exit, and on_update logic and we'll do the rest" approach relieves a lot of complexity, boilerplate, and footguns. It also provides a standard structure in the schedule that other plugins can make assumptions about.

Maybe we should support both?

Given that we're talking about footguns / ergonomics / api surface, can you port the app setup of the current `state.rs` to the proposed approach (with functional equivalency ... including command buffer application) so we can do a comparison?

```rust
fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_state(AppState::Menu)
        .add_startup_system(setup)
        .add_system_set(SystemSet::on_enter(AppState::Menu).with_system(setup_menu))
        .add_system_set(SystemSet::on_update(AppState::Menu).with_system(menu))
        .add_system_set(SystemSet::on_exit(AppState::Menu).with_system(cleanup_menu))
        .add_system_set(SystemSet::on_enter(AppState::InGame).with_system(setup_game))
        .add_system_set(
            SystemSet::on_update(AppState::InGame)
                .with_system(movement)
                .with_system(change_color),
        )
        .run();
}
```

Reply:
> If you don't buy that "state update logic" should be named / addressable / orderable, then the symmetry argument starts to break down.

I don't think I disagree with *this*.

> I think standardizing "state lifecycles" for the common case is "nice".

> I think it should be a named group of systems that exists within the larger CoreSystems::Update set.

But I'm skeptical about this. I guess I just worry that a default location for these systems is impossible to get right.

> Ahh so this is fully an exercise for the user. Not something with a standard place in the schedule. Not yet sure how I feel about that. Seems like a footgun / something easy to forget.

We did try to make it as friendly as possible by providing an `apply_state_transition::<S: State>` system counterpart to `apply_system_buffers`. However, I don't think the general "these must run in an exclusive system" nature of the design is negotiable.

There's one or two factors contributing to this.

1. In our understanding, run criteria that mutate data or loop need to go, and the only way we've found to avoid them is by encapsulating loops and other "spliced-in" behavior *inside* exclusive systems. That way, executing a graph is always a simple pass where each system runs at most once (which I think will be essential for `foo.after_commands(bar)`, etc.).
2. Every node must have a unique name. Since system sets are basically *just* names, you couldn't have `OnEnter(X)` show up multiple times in a single graph. (Even before that, you can't clone systems.)

> Maybe we should support both?

I don't think I'm staunchly against this, but I think it would be really important to identify the 5% where `OnUpdate(X)` maybe isn't the right solution.

> Can you port the app setup of the current `state.rs` to the proposed approach?

(Will do, consider this a temporary placeholder for now.)

Reply:
Could this just be `apply_state_transition.after(CoreSystems::Update)`?

Reply:
Here is my proposal for states (accounting for the Systems->Schedules changes [I proposed here](https://github.com/bevyengine/rfcs/pull/45#discussion_r953198893)):

```rust
fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
         // init_state adds all state resources, sets, and schedules ... deriving the State trait supplies the
         // list of possible enum values, which removes the need for users to manually supply each value.
         // This adds OnEnter/OnExit schedules and OnUpdate sets for each state value.
         // 
         // This also (by default) calls:
         // `app.add_system(apply_state_transition::<AppState>.in_set(CoreSystems::PreUpdateStateTransition))`
         // which ensures that state transitions happen _before_ update logic, ensuring that for a given
         // OnEnter(State), "state update logic" runs _after_ it. This is important, because OnEnter defines
         // "init logic" for the state. Update logic can and should rely on OnEnter logic being called first,
         // and OnUpdate should run at least once after OnEnter before rendering the frame.
         // CoreSystems::PreUpdateStateTransition would also apply Commands after that set, avoiding the foot-gun
         // of users forgetting to apply their spawn / despawn Commands for OnEnter / OnExit
         // Ex: OnEnter spawns a thing, OnUpdate properly positions the thing
         //
         // It also calls:
        //`app.configure_set(OnUpdate(AppState::Menu)
        //    .in_set(CoreSystems::Update).run_if(state_equals(AppState::Menu)))`
        //  For each state, which ensures the "update" set runs during the "update" phase of the app when appropriate
        .init_state::<AppState>()
        .add_system(setup.in_schedule(Startup))
        .add_system(setup_menu.in_schedule(OnEnter(AppState::Menu)))
        .add_system(menu.in_set(OnUpdate(AppState::Menu)))
        .add_system(cleanup_menu.in_schedule(OnExit(AppState::Menu)))
        .add_system(setup_game.in_schedule(OnEnter(AppState::InGame)))
        .add_systems(
            systems![movement, change_color]
                .in_set(OnUpdate(AppState::InGame))
         )
        .run();
}
```

In my mind, this is _much_ better ux. It reads better. It requires less typing. It requires fewer user decisions by default (where to apply commands, where to apply state transitions), it has symmetry across state events, and it _still_ allows users to do their own thing manually, as needed (using the `run_if` and `apply_state_transitions` pieces already defined in this RFC).

This "standard pattern" fits _every_ state-driven app I've ever written, and I suspect that will be true for basically every Bevy app. And again, for the special cases where it isn't perfect, in most cases users can just add their additional logic on top using run_if / apply_state_transitions.

And in (what I expect to be) the _very rare_ event that the defaults aren't right for a users' app, we can have a variant of `init_state::<AppState>` that doesn't add the defaults, such as `init_state_manual::<AppState>` or `add_plugin(StatePlugin::<AppState>::manual())`.

> (like the after_all I added here).

Not a huge fan of "after_all" as a concept, as that can only be true for the first caller of `after_all()`. Seems like we just need to add the appropriate ordering constraints. `apply_state_transition::<AppState>.in_set(CoreSystems::ApplyUpdateCommands).after(CoreSystems::Update).before(CoreSystems::PostUpdate)`

Or alternatively, have a well defined point in the schedule for state transitions so users don't need to think about this by default :)

Reply:
I *really* like Cart's suggestion above. Unless @maniwani or others have objections; I'm going to add this to the RFC in the next day or two.

Reply:
> Next day or two.

Please wait. I'd like a chance to respond.

Reply:
I'm fine with `OnUpdate(X)` if you think the practicality outweighs any cons. I've said all I had to say about discoverability.

> Seems like we just need to add the appropriate ordering constraints. `apply_state_transition::<AppState>.in_set(CoreSystems::ApplyUpdateCommands).after(CoreSystems::Update).before(CoreSystems::PostUpdate)`

This is more what I was getting at. I'd like you to answer a couple questions before we update the RFC (along with method name bikeshedding in the other comment thread).

- Would many types of `apply_state_transition::<S>` systems go in `CoreSystems::PreUpdateStateTransition`?
- If so, does the relative execution order between them not matter? (Or is that a convention we need to adopt?)

Reply:
> I'm fine with OnUpdate(X) if you think the practicality outweighs any cons. I've said all I had to say about discoverability.

Awesome. I believe in it enough to bet on it / im willing to eat my words / admit I was wrong if we discover users don't like it.

> Would many types of apply_state_transition::<S> systems go in CoreSystems::PreUpdateStateTransition? If so, does the relative execution order between them not matter? Is that a convention we need to adopt? 

Yup that is what I was suggesting. But thats a good point. Maybe we should make the set generic on the state type so that it is addressable / orderable?